# Размещение баннеров

Представим, что международное круизное агентство Carnival Cruise Line решило себя разрекламировать с помощью баннеров и обратилось для этого к вам. Чтобы протестировать, велика ли от таких баннеров польза, их будет размещено всего 20 штук по всему миру. Вам надо выбрать 20 таких локаций для размещения, чтобы польза была большой и агентство продолжило с вами сотрудничать.

Агентство крупное, и у него есть несколько офисов по всему миру. Вблизи этих офисов оно и хочет разместить баннеры — легче договариваться и проверять результат. Также эти места должны быть популярны среди туристов.

In [1]:
import numpy as np
import pandas as pd
from sklearn.cluster import MeanShift
from collections import Counter

Для поиска оптимальных мест воспользуемся базой данных крупнейшей социальной сети, основанной на локациях — Foursquare.

Часть открытых данных есть, например, на сайте archive.org:

https://archive.org/details/201309_foursquare_dataset_umn

Скачайте любым удобным образом архив fsq.zip с этой страницы.

Нас будет интересовать файл checkins.dat.

Для удобной работы с этим документом преобразуем его к формату csv, удалив строки, не содержащие координат — они неинформативны для нас:

In [2]:
#Reading the data file
checkins = pd.read_csv('checkins.dat', sep='|', skipinitialspace=True, skiprows=[1], low_memory=False)
print(checkins.info())
checkins.head()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1021967 entries, 0 to 1021966
Data columns (total 6 columns):
 #   Column            Non-Null Count    Dtype  
---  ------            --------------    -----  
 0   id                1021967 non-null  object 
 1   user_id           1021966 non-null  float64
 2   venue_id          1021966 non-null  float64
 3   latitude          396634 non-null   float64
 4   longitude         396634 non-null   float64
 5   created_at        1021966 non-null  object 
dtypes: float64(4), object(2)
memory usage: 46.8+ MB
None


id       user_id   venue_id   latitude        longitude       \
0  984301   2041916.0     5222.0             NaN             NaN   
1  984222     15824.0     5222.0       38.895112      -77.036366   
2  984315   1764391.0     5222.0             NaN             NaN   
3  984234     44652.0     5222.0       33.800745      -84.410520   
4  984249   2146840.0     5222.0             NaN             NaN   

      created_at        
0  2012-04-21 17:39:01  
1  2012-04-21 17:43:47  
2  2012-04-21 17:37:18  
3  2012-04-21 17:43:43  
4  2012-04-21 17:42:58

In [3]:
#Stripping columns names (deleting whitespaces)
checkins.columns = checkins.columns.str.strip()
#Dropping rows with NaN values
checkins = checkins.dropna()
print('Shape without NaN: %s' % str(checkins.shape))
checkins.head()

Shape without NaN: (396634, 6)


id    user_id  venue_id   latitude   longitude           created_at
1   984222     15824.0    5222.0  38.895112  -77.036366  2012-04-21 17:43:47
3   984234     44652.0    5222.0  33.800745  -84.410520  2012-04-21 17:43:43
7   984291    105054.0    5222.0  45.523452 -122.676207  2012-04-21 17:39:22
9   984318   2146539.0    5222.0  40.764462 -111.904565  2012-04-21 17:35:46
10  984232     93870.0  380645.0  33.448377 -112.074037  2012-04-21 17:38:18

In [4]:
checkins.info()

<class 'pandas.core.frame.DataFrame'>
Index: 396634 entries, 1 to 1021964
Data columns (total 6 columns):
 #   Column      Non-Null Count   Dtype  
---  ------      --------------   -----  
 0   id          396634 non-null  object 
 1   user_id     396634 non-null  float64
 2   venue_id    396634 non-null  float64
 3   latitude    396634 non-null  float64
 4   longitude   396634 non-null  float64
 5   created_at  396634 non-null  object 
dtypes: float64(4), object(2)
memory usage: 21.2+ MB


Теперь необходимо кластеризовать данные координаты, чтобы выявить центры скоплений туристов. Поскольку баннеры имеют сравнительно небольшую площадь действия, нам нужен алгоритм, позволяющий ограничить размер кластера и не зависящий от количества кластеров.

Эта задача — хороший повод познакомиться с алгоритмом MeanShift. Его описание при желании можно посмотреть в sklearn user guide. Используйте MeanShift, указав bandwidth=0.1, что в переводе из градусов в метры колеблется примерно от 5 до 10 км в средних широтах.

Примечание: на 396634 строках кластеризация будет работать долго. Быть очень терпеливым не возбраняется — результат от этого только улучшится. Но для того, чтобы сдать задание, понадобится сабсет из первых 100 тысяч строк. Это компромисс между качеством и затраченным временем. Обучение алгоритма на всём датасете занимает около часа, а на 100 тыс. строк — примерно 2 минуты, однако этого достаточно для получения корректных результатов.

In [5]:
subset = checkins.head(100000)
ms = MeanShift(bandwidth=0.1)
ms.fit(subset[['latitude', 'longitude']])

MeanShift(bandwidth=0.1)

In [6]:
labels = ms.labels_
cluster_centers = ms.cluster_centers_

Некоторые из получившихся кластеров содержат слишком мало точек — такие кластеры не интересны рекламодателям. Поэтому надо определить, какие из кластеров содержат, скажем, больше 15 элементов. Центры этих кластеров и являются оптимальными для размещения.

In [7]:
label_counts = Counter(labels)

filtered_clusters = {label: count for label, count in label_counts.items() if count > 15}

filtered_centroids = [cluster_centers[label] for label in filtered_clusters]

Как мы помним, 20 баннеров надо разместить близ офисов компании. Найдем на Google Maps по запросу Carnival Cruise Line адреса всех офисов:

33.751277, -118.188740 (Los Angeles)

25.867736, -80.324116 (Miami)

51.503016, -0.075479 (London)

52.378894, 4.885084 (Amsterdam)

39.366487, 117.036146 (Beijing)

-33.868457, 151.205134 (Sydney)

Осталось определить 20 ближайших к ним центров кластеров. Т.е. посчитать дистанцию до ближайшего офиса для каждой точки и выбрать 20 с наименьшим значением.

In [8]:
def distance(x, y):
    return np.sqrt(np.sum((x - y)**2))

offices = [[33.751277, -118.188740],
           [25.867736, -80.324116],
           [51.503016, -0.075479],
           [52.378894, 4.885084],
           [39.366487, 117.036146],
           [-33.868457, 151.205134]]

centroids_distance = []
for office in offices:
    for center in filtered_centroids:
        centroids_distance.append([center.tolist(), distance(center, office)])

closest_centroids = sorted(centroids_distance, key=lambda x: x[1])[:20]

print(closest_centroids)

[[[-33.86063042857143, 151.20477592857142], 0.007834758163109155], [[52.37296399032261, 4.892317222580647], 0.009353316185992226], [[25.845672264285717, -80.31889059642857], 0.022674066158368207], [[51.502991260887086, -0.12553728870967767], 0.05005829482278787], [[33.8098779552631, -118.1489238069082], 0.07084773242715979], [[25.7858124199675, -80.2179380368254], 0.13410903336184654], [[25.705349721052595, -80.28342873815798], 0.16740596425033946], [[26.010098249285683, -80.19999058571432], 0.18887596060185083], [[33.888325342758606, -118.04892817172427], 0.19577945647764128], [[33.87298601157018, -118.36209114655645], 0.21181053682436798], [[33.97257482142857, -118.16837066666665], 0.22223329073178233], [[26.138843786842077, -80.33434683684207], 0.2713007595066735], [[33.98393587403844, -118.00740497307687], 0.29497888680046563], [[26.120862658633104, -80.1589066802157], 0.3022701186924605], [[33.81730643390889, -117.891249170958], 0.30473050307840693], [[34.06039755458241, -118.2487

Выберите из получившихся 20 центров тот, который наименее удален от ближайшего к нему офиса. Ответ в этом задании — широта и долгота этого центра, записанные через пробел.

In [9]:
print('Minimal distance: %f' % closest_centroids[0][1])
print('Coordinates of closest centroid: %s' % str(closest_centroids[0][0]))

Minimal distance: 0.007835
Coordinates of closest centroid: [-33.86063042857143, 151.20477592857142]
